feat(gateway): always-on base request log with five opt-in capture settings (#6921)
unkeyed/unkey: docs/product/platform/gateway/policies/logging.mdx

---
title: Logging
description: "Control what the gateway stores in the request log. Add headers and bodies to the basic log entries for debugging."
---

import DeployBeta from "/snippets/deploy-beta.mdx";

<DeployBeta />

> The gateway always records a basic log entry for each request. The dashboard uses these entries for the traffic and latency charts. You cannot turn them off.

Reviewer: No need to call this out specifically. I think it's clear already.
```suggestion
The gateway always records a basic log entry for each request. The dashboard uses these entries for the traffic and latency charts.
```

> The logging policy adds sensitive data to the log entries of matched requests: headers and bodies. Each capture is a separate opt-in setting. Without an enabled logging policy, the gateway stores only the basic entries.

Reviewer: Make it sound less scary

```suggestion
The logging policy can add potentially sensitive data to the log entries of matched requests: headers and bodies. You can opt into storing bodies or headers individually.
```

To capture data for all traffic, add a logging policy with no match conditions. To capture data only for specific routes, add match conditions to the policy.

## Always logged

The gateway records this data for every request:

- HTTP method, host, and path
- Status code
- Request ID and timestamp
- Deployment and instance identifiers
- Region
- Latency breakdown

| Metric           | Description                                                |
| ---------------- | ---------------------------------------------------------- |
| Total latency    | End-to-end time from when the gateway received the request |
| Instance latency | Time your app spent processing the request                 |
| Gateway latency  | Overhead added by the gateway (policy evaluation, routing) |

## Opt-in capture

A logging policy has four independent settings. Each one adds data to the log entries of matched requests:

| Setting          | Captured data                                                                    |
| ---------------- | -------------------------------------------------------------------------------- |

> | Request headers  | Request headers, query string, query parameters, user agent, and client IP       |

Reviewer: query string?
that's not a header... we should probably have a separate flag for that

| Response headers | Response headers                                                                  |
| Request body     | Request body, up to the capture limit                                             |
| Response body    | Response body, up to the capture limit                                            |

The request headers setting includes query data because URLs can contain secrets, for example `?api_key=...`. It includes the user agent and client IP because they identify the client.

> If more than one enabled logging policy matches a request, the gateway combines their settings.

Reviewer: we should probably just prevent having 2 logging policies

Reviewer: nvm, then you couldn't log headers on one path and bodies on another..

makes me wonder though if we're overcomplicating it.

Author: :sweat_smile: im not sure it does make sense in my head to only log some things in one path and more in others  [](edited-with-Axolo)

Reviewer: it definitely makes sense
I'm just trying to gauge where we want to position ourselves between simplicity that's easy to configure and full control with a lot of options [](edited-with-Axolo)

## Retention

> Unkey keeps logs for 30 days.

Reviewer: This is not true, it depends on your plan: https://www.unkey.com/docs/platform/workspaces/quotas#log-retention
